File: java/L000/LeetCode0045.java
package L000;

public class LeetCode0045 {
    public int jump(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return 0;
        }
        int[] steps = new int[nums.length];
        steps[nums.length - 1] = 0;
        for (int i = nums.length - 2;i >= 0;--i) {
            if (i + nums[i] >= nums.length - 1) {
                steps[i] = 1;
            } else {
                int step = nums.length;
                for (int j = nums[i];j >= 1;--j) {
                    step = Math.min(step, 1 + steps[i + j]);
                }
                steps[i] = step;
            }
        }
        return steps[0];
    }
}

File: cpp/L000/LeetCode0045.cc
#include <limits>
#include <vector>
using std::vector;

class Solution {
public:
    int jump(vector<int>& nums) {
        int N = nums.size();
        vector<int> dp(N, std::numeric_limits<int>::max());
        dp[0] = 0;
        for (int i = 0;i < N - 1;++i) {
            for (int j = 1;j <= nums[i] && i + j < N;++j) {
                dp[i + j] = std::min(dp[i + j], dp[i] + 1);
            }
        }
        return dp[N - 1];
    }
};